src/fix_ea/fix_ea.c

```diff
@@ -0,0 +1,555 @@
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+ * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <sys/types.h>
+#include <sys/queue.h>
+#include <sys/stat.h>
+#include <sys/extattr.h>
+#include <fcntl.h>
+#include <fts.h>
+#include <libgen.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sysexits.h>
+#include <unistd.h>
+#include <err.h>
+
+#define EX_EA_CORRUPTED         1
+
+#define F_NONE                  0x0000
+#define F_APPEND_NULL_ALL       0x0001
+#define F_CHECK_AFP_EA          0x0002
+#define F_DRY_RUN               0x0004
+#define F_FIX_AFP_EA            0x0008
+#define F_APPEND_NULL           0x0010
+#define F_VERBOSE               0x0020
+#define F_DEBUG                 0x0040
+#define F_RECURSIVE             0x0080
+
+#define AFP_EA_CORRUPTED(v)     (v[0] == 0 && v[1] == 'F' && v[2] == 'P')
+
+
+struct xattr {
+	char *name;
+	char *value;
+	size_t length;
+	TAILQ_ENTRY(xattr) link;
+	TAILQ_ENTRY(xattr) afp_link;
+	TAILQ_ENTRY(xattr) append_link;
+};
+
+TAILQ_HEAD(xattr_list, xattr);
+
+
+void
+usage(const char *path)
+{
+	fprintf(stderr,
+		"Usage: %s [OPTIONS] <path|file>\n"
+		"Where option is:\n"
+		"     -a                # append null byte to all extended attributes\n"
+		"     -c                # check if AFP extended attributes are corrupted\n"
+		"     -C                # dry run (no changes are made)\n"
+		"     -d                # debug mode\n"
+		"     -f                # fix AFP extended attributes\n"
+		"     -n <EA>           # append null byte\n"
+		"     -r                # recursive\n"
+		"     -v                # verbose\n\n"
+		"Exit codes:\n"
+		"      1 if corrupted\n"
+		"      0 if not corrupted or fixed\n",
+		path
+	);
+
+	exit(EX_USAGE);
+}
+
+static int
+get_extended_attributes(int fd, struct xattr_list *xlist)
+{
+	char *buf;
+	int i, ch, ret, buflen;
+
+	if ((ret = extattr_list_fd(fd, EXTATTR_NAMESPACE_USER, NULL, 0)) < 0)
+		return (EX_OK);
+
+	if ((buf = malloc(ret)) == NULL) {
+		warn("malloc");
+		return (-1);
+	}
+
+	buflen = ret;
+	if ((ret = extattr_list_fd(fd, EXTATTR_NAMESPACE_USER, buf, buflen)) < 0) {
+		free(buf);
+		warn("extattr_list_fd");
+		return (-1);
+	}
+
+	for (i = 0;i < ret;i += ch + 1) {
+		struct xattr *xptr = NULL;
+		char *name, *value;
+		int getret;
+
+		ch = (unsigned char)buf[i];
+		if ((name = malloc(ch)) == NULL) {
+			err(EX_OSERR, NULL);
+			continue;
+		}
+
+		strncpy(name, &buf[i + 1], ch);
+		name[ch] = '\0';
+
+		if (strncmp(name, "DosStream.", 10) != 0) {
+			free(name);
```

Comment:
As this is a configurable parameter to the `vfs_streams_xattr` it could be handy to be able to redefine streams prefix name from the command line. It's unlikely that it was altered, but...

Reply:
I'm not very concerned about this. This tool is for a specific problem. If we encounter this in the wild, then that might change.